crates/doc/src/inference.rs

```diff
@@ -1484,6 +1561,152 @@ impl Shape {
+                        ..Default::default()
+                    }))
+                }
+
+                let hint = self.object.widen::<N>(fields, loc, is_first_time);
+
+                match (hint, loc) {
+                    (true, _) => true,
+                    (false, Location::Root) => self.object.properties.len() > MAX_ROOT_FIELDS,
+                    (false, _) => self.object.properties.len() > MAX_NESTED_FIELDS,
+                }
+            }
+
+            crate::Node::Array(arr) => {
+                let mut shape = self.array.additional.take().unwrap_or(Box::new(Shape {
+                    // Start out maximally constrained
+                    type_: types::INVALID,
+                    provenance: Provenance::Inline,
+                    ..Default::default()
+                }));
+
+                // Look at each element in the observed array and widen the shape to accept it
+                let hint = arr.iter().enumerate().fold(false, |accum, (idx, node)| {
+                    accum || shape.widen(node, loc.push_item(idx))
+                });
+
+                self.array.additional = Some(shape);
+                self.type_ = self.type_ | types::ARRAY;
+
+                hint
+            }
+            crate::Node::Bool(_) => {
+                self.type_ = self.type_ | types::BOOLEAN;
+                false
+            }
+            crate::Node::Bytes(_) => {
+                self.type_ = self.type_ | types::STRING;
+
+                let partial_stringshape = StringShape {
+                    content_encoding: Some("base64".to_string()),
+                    ..Default::default()
+                };
+
+                self.string = StringShape::union(self.string.clone(), partial_stringshape);
+                false
+            }
+            crate::Node::Null => {
+                self.type_ = self.type_ | types::NULL;
+                false
+            }
+            crate::Node::Number(_num) => {
+                self.type_ = self.type_ | types::INT_OR_FRAC;
+                false
+            }
+            crate::Node::String(_s) => {
+                self.type_ = self.type_ | types::STRING;
+                let partial_stringshape = StringShape {
+                    // TODO: implement format widening
+                    ..Default::default()
+                };
+
+                self.string = StringShape::union(self.string.clone(), partial_stringshape);
+
+                false
+            }
+        }
+    }
+
+    // Prune any locations in this shape that have more than the allowed fields,
+    // squashing those fields into the `additionalProperties` subschema for that location.
+    pub fn enforce_field_count_limits(&mut self, loc: Location) {
+        // TODO: If we implement inference/widening of array tuple shapes
+        // then we'll need to also check that those aren't excessively large.
+        if !self.type_.overlaps(types::OBJECT) {
```

Comment:
you still need to recurse for cases like `{"arr": [{"so":1, "many":2, "props":3}]}`

Reply:
Good point!